src/hotspot/share/opto/type.cpp

```diff
@@ -1744,6 +1744,22 @@ const Type *TypeInt::narrow( const Type *old ) const {
   return this;
 }
 
+bool TypeInt::can_overflow(int opcode, const TypeInteger* other) const {
+  const TypeInt* other_int = other->isa_int();
+  if (other_int == nullptr || empty() || other->empty()) {
+    return true;
+  }
+  switch (opcode) {
+  case Op_AddI:
+    return add_underflows(_lo, other_int->_lo) ||
+           add_overflows(_hi, other_int->_hi);
+  case Op_SubI:
+    return subtract_underflows(_lo, other_int->_hi) ||
+           subtract_overflows(_hi, other_int->_lo);
+  }
+  return true;
+}
```

Comment:
Do you maybe want to assert that no other opcode comes in? Or is there a need for non add/sub opcodes to be passed in?

Reply:
Added assertions